### ISSUE TITLE
Consul ACL Token does not rotate on allocations when changed in agent config

### ISSUE THREAD

**Issue by artyomturkin** [hashicorp/nomad#9813]
### Nomad version
v1.0.0

### Operating system and Environment details
Ubuntu 20.04

### Issue
If Consul ACL Token is removed from Consul (using Vault to generate them for example). Allocations fail to register in consul. There are no error logs in nomad output on client. In Consul logs there are entries that show attempts to register service without a proper token (accessorID is empty in log)

### Reproduction steps
1. Start nomad with consul token
2. Start nomad job with service registration
3. Remove token from consul
4. Replace consul token with new one in nomad config
5. Restart nomad agent
6. Republish nomad job to fix the issue

**Comment by krishicks** [comment]
I was able to confirm this behavior with [vagrant-nomad](https://github.com/krishicks/vagrant-nomad).


```
git clone git@github.com:krishicks/vagrant-nomad
cd vagrant-nomad
vagrant up

# wait for Nomad server/client to be healthy in Consul

./add-consul-acl.sh

# with the Consul token emitted above
export CONSUL_HTTP_TOKEN=..
export CONSUL_HTTP_ADDR=http://10.199.0.10:8500
export NOMAD_ADDR=http://10.199.0.10:4646

consul intention create count-dashboard count-api
nomad job run countdash.nomad

# wait for count-dashboard and count-api to become healthy in Consul

# delete old token
id="$(consul acl token list -format json | jq -r 'map(select(.Description | contains("nomad-client-00")) | .AccessorID) | .[]')"
consul acl token delete -id $id

# add new token
agentToken="$(consul acl token create -description "New Nomad agent token" -policy-name "nomad-client" -format json | jq -r '.SecretID')"
cmd="$(cat <<EOF
sudo sed -i 's/consul {.*}/consul { token = \"$agentToken\" }/g' /opt/nomad/config.hcl
sudo systemctl restart nomad
EOF
)"
vagrant ssh nomad-client-00 -c "$cmd" >/dev/null 2>&1
```

After the above steps, we see errors in Consul not just regarding the task registration, but a bunch of other issues:

```
Feb 10 00:05:44 ubuntu-focal consul[2549]:     2021-02-10T00:05:44.458Z [ERROR] agent.client: RPC failed to server: method=Catalog.Register server=10.199.0.10:8300 error="rpc error making call: ACL not found"
Feb 10 00:05:44 ubuntu-focal consul[2549]:     2021-02-10T00:05:44.458Z [WARN]  agent: Service registration blocked by ACLs: service=_nomad-task-719c37be-5205-8a6f-c37e-ce76fa6d1904-group-dashboard-count-dashboard-9002 accessorID=
...
Feb 10 00:06:32 ubuntu-focal consul[2549]:     2021-02-10T00:06:32.954Z [ERROR] agent.client: RPC failed to server: method=Intention.Match server=10.199.0.10:8300 error="rpc error making call: ACL not found"
Feb 10 00:06:47 ubuntu-focal consul[2549]:     2021-02-10T00:06:47.619Z [ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=10.199.0.10:8300 error="rpc error making call: ACL not found"
```

After getting into this state I was unable to get things fully happy again. The task is able to be registered, but the other RPC errors remain:

```
Feb 10 00:15:11 ubuntu-focal consul[2549]:     2021-02-10T00:15:11.423Z [INFO]  agent: Synced service: service=_nomad-client-n44iwtmbke7323p6lqb4tzvn2okxcwjg
Feb 10 00:15:11 ubuntu-focal consul[2549]:     2021-02-10T00:15:11.430Z [INFO]  agent: Synced service: service=_nomad-task-d3ffa50a-06d7-0cee-c54b-b39da2741b1a-group-dashboard-count-dashboard-9002
Feb 10 00:15:11 ubuntu-focal consul[2549]:     2021-02-10T00:15:11.438Z [INFO]  agent: Synced service: service=_nomad-task-d3ffa50a-06d7-0cee-c54b-b39da2741b1a-group-dashboard-count-dashboard-9002-sidecar-proxy
Feb 10 00:15:11 ubuntu-focal consul[2549]:     2021-02-10T00:15:11.447Z [INFO]  agent: Synced service: service=_nomad-task-1ea9333e-922c-2ed5-ac25-397a964b844e-group-api-count-api-9001
Feb 10 00:15:11 ubuntu-focal consul[2549]:     2021-02-10T00:15:11.451Z [INFO]  agent: Synced service: service=_nomad-task-1ea9333e-922c-2ed5-ac25-397a964b844e-group-api-count-api-9001-sidecar-proxy
Feb 10 00:15:12 ubuntu-focal consul[2549]:     2021-02-10T00:15:12.445Z [ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=10.199.0.10:8300 error="rpc error making call: ACL not found"
Feb 10 00:15:13 ubuntu-focal consul[2549]:     2021-02-10T00:15:13.838Z [ERROR] agent.client: RPC failed to server: method=Intention.Match server=10.199.0.10:8300 error="rpc error making call: ACL not found"
```

**Comment by manveru** [comment]
So since I've been fighting this for the last few months, here's my temporary solution until this gets fixed. 
I don't pass a Consul token to Nomad anymore, and rely instead on the default token set in the Consul configuration, which is much easier to rotate.
Unfortunately this might not work for all people, depending on your security requirements.

**Comment by mwild1** [comment]
I've seen using the default token suggested a number of times. Correct me if I'm wrong, but this essentially allows unauthenticated users to perform any actions the default token can (which in this case means at least managing services).

It seems to me that this is quite a backwards step for security in most deployments that are using ACLs in the first place.

**Comment by manveru** [comment]
Yes, it's less than ideal, but there was no other way I could find to actually get my services to (de)register after the token TTL has passed.
I'd be happy for any alternatives, but can't find any other than killing all Nomad allocations after their token expires and cause a ton of chaos (and even then, it'd just leave service entry zombies because deregistration fails).

The other approach would be to have tokens that just don't expire, which is arguably worse in some ways, since you have no way of knowing what tokens are still in use, and if you leak one it'll be quite damaging.

We only deploy software we build ourselves from scratch and generally trust our developers that have access to deploy nomad jobs, we also don't store anything in Consul KV and rely on Vault for that, so the worst that someone could do is to mess with service registrations.

To be honest, at this point I doubt anyone is seriously doing token and cert rotation in production the way HashiCorp recommends, and probably also they don't do it themselves, otherwise issues like these wouldn't pop up so often across all their products.

**Comment by johnalotoski** [comment]
We are seeing this issue in a Nomad cluster with ACLs and Connect jobs (Nomad: 1.1.4, Consul 1.10.2, Vault 1.8.2).  Vault is managing consul and nomad tokens: expiring/renewing leases, creating new tokens as needed.  We have token rotation set up such that vault agent template config is used to write out new consul and nomad tokens to config files on a timely basis and then consul and nomad services are restarted and consume those config files with the newly rotated tokens.

This issue can be triggered in the following manner:

1) Start a new Nomad Connect job

2) Check Consul and Nomad logs (client and server nodes) after job start up -- all looks good

3) Force a manual token rotation by restarting vault-agent on the Nomad client and re-generating the Consul template files with Consul and Nomad tokens, thereby also restarting Nomad and Consul with the new rotated token configuration for consul default token and consul agent token

4) Check Consul and Nomad logs again (client and server nodes) -- all looks good

5) Allow "X" period of time to go by until the original tokens used in step (1) expire.  We can see the expiration of these tokens from the logs on the Vault server service, example:
```
2021-09-29T09:59:10.620Z [INFO] expiration: revoked lease: lease_id=pki/issue/client/$LEASE_ID
2021-09-29T09:59:09.152Z [INFO] expiration: revoked lease: lease_id=auth/aws/login/$LEASE_ID
2021-09-29T09:59:09.108Z [INFO] expiration: revoked lease: lease_id=consul/creds/consul-agent/$LEASE_ID
2021-09-29T09:59:09.055Z [INFO] expiration: revoked lease: lease_id=consul/creds/consul-default/$LEASE_ID
```
Immediately, the Nomad client server with the $LEASE_ID corresponding to the `consul/creds/consul-agent` token, used in the consul agent config stanza of `{ "acl": { "tokens": { "agent": "$AGENT_TOKEN" } } }` begins throwing consul service log warnings and errors similar to those seen above:
```
2021-09-29T18:40:31.841Z [ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=172.16.0.10:8300 error="rpc error making call: ACL not found"
2021-09-29T18:40:03.487Z [WARN] agent.cache: handling error in Cache.Notify: cache-type=intention-match error="rpc error making call: ACL not found" index=1577791
2021-09-29T18:40:03.487Z [ERROR] agent.client: RPC failed to server: method=Intention.Match server=172.16.1.10:8300 error="rpc error making call: ACL not found"
2021-09-29T18:38:51.243Z [WARN] agent.cache: handling error in Cache.Notify: cache-type=intention-match error="rpc error making call: rpc error making call: ACL not found" index=1577791
2021-09-29T18:38:51.243Z [ERROR] agent.client: RPC failed to server: method=Intention.Match server=172.16.2.10:8300 error="rpc error making call: rpc error making call: ACL not found"
2021-09-29T18:38:39.510Z [WARN] agent.cache: handling error in Cache.Notify: cache-type=resolved-service-config error="rpc error making call: ACL not found" index=1577791
2021-09-29T18:38:39.510Z [ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=172.16.0.10:8300 error="rpc error making call: ACL not found"
2021-09-29T18:37:19.447Z [WARN] agent.cache: handling error in Cache.Notify: cache-type=intention-match error="rpc error making call: rpc error making call: ACL not found" index=1577791
```
Interestingly, the expiring lease for the token with the `consul/creds/consul-default` role does not appear to cause any issue.  Only the consul agent token rotation is a problem, even though the consul agent token is actually also holding the same consul default token policy as the consul default token itself.

6) To try and fix this state from the errors and warning being shown in the logs, some combination of purging the Nomad job, restarting consul and/or nomad and redeploying the Nomad jobs afresh appears to be required.  Just restarting Consul and/or Nomad jobs alone is insufficient.  Neither does the `consul acl set-agent-token agent` CLI command help.

Unless I've misunderstood the problem, overlooked a work-around or a different approach should be taken to token rotation, this problem appears to make successful token rotation just outright impossible at the moment.

**Comment by ryanm-sq** [comment]
Adding in an additional data point since we're seeing this issue as well. It seems that services and checks get "cached" to `/opt/consul/[services|checks]` (Ubuntu 20.04) and contain the last consul token used. I believe the token written in the cached check configs get used to sync services/checks and once the token expires, will result in the error described here.

These files do not appear to refresh automatically with a restart/reload of the consul agent when tokens are auto-rotated. I can temporarily resolve the issue by clearing those directories and forcing a reload of the consul agent but the issue comes back once the new token expires.

**Comment by tgross** [comment]
Starting in [Nomad 1.7.0-beta.1](https://github.com/hashicorp/nomad/releases/tag/v1.7.0-beta.1) we've deprecated the use of Consul tokens in the Nomad agent configuration for purposes of giving workloads access to Consul. Nomad will use workload identities to sign into Consul for purposes of getting Consul tokens for those workloads, rather than using the agent's token.

**Comment by Legogris** [comment]
@tgross As users, this thread is the first time we hear of this deprecation and I can't seem  to find anything about it in [1.7.0 release notes](https://github.com/hashicorp/nomad/releases/tag/v1.7.0). It is noted in the upgrade guide, but we're still on 1.6 due to a "stay one behind current for stability" purposes.

Deprecating the de-facto consul-nomad integration is a quite significant change and looking at open issues, not everyone can be expected to start migrating to workload identities just yet...

In the meantime, are availability-impacting bugs in the now deprecated integration not something that can be addressed for existing users on 1.6 relying on it?

**Comment by tgross** [comment]
> Deprecating the de-facto consul-nomad integration is a quite significant change and looking at open issues, not everyone can be expected to start migrating to workload identities just yet...

Of course! Which is why the [upgrade guide for 1.7.0](https://developer.hashicorp.com/nomad/docs/upgrade/upgrade-specific#nomad-1-7-0) says that it won't be removed until Nomad 1.9.

> In the meantime, are availability-impacting bugs in the now deprecated integration not something that can be addressed for existing users on 1.6 relying on it?

Not at all. But this issue is about config reloads, which _by design_ don't impact running allocations (as much as it's possible). If you've got an issue impacting availability, please open one!

**Comment by Legogris** [comment]
> Not at all. But this issue is about config reloads, which by design don't impact running allocations (as much as it's possible). If you've got an issue impacting availability, please open one!

I'm not sure I follow. As this issue affects Nomad clients ability to register/deregister services for their jobs in Consul, it manifests as unavailability for those services.

This issue is also related to (and hairier to work around because of) these:
- #4593
- #3885
- #20159

**Comment by Legogris** [comment]
I could see how this issue might be closed as a duplicate of an already open one - my concern here is about the "nofix" statement, indicating that any fix for these issues should not be expected for 1.6~1.8, either.

**Comment by tgross** [comment]
@Legogris if you've got an issue impacting you, please just open a new issue.

**Comment by Himura2la** [comment]
We experience this or similar issue on Nomad 1.9.1 with Workload Identities. We rotated the token used by Nomad clients while migrating to WI, maybe that's the source of the issue, IDK.
I had a hard time forcing Nomad and Consul to forget about old tokens...

Also, we experience the issue only on two jobs (in two datacenters), which is pretty strange, since they are no different from other dozens of jobs...
